### PR TITLE
Add constructors for `device_buffer` that take an explicit alignment parameter

### LINKED ISSUE #2329: [FEA] Support device buffer allocations with specific alignment
**Is your feature request related to a problem? Please describe.**

The MR interface allows me to request an alignment of the returned pointer (ok, it's currently ignored and device MRs always return 256-byte aligned allocations, but that may change).

However, if I have a device_buffer I can't request an alignment. This is problematic because if I want an untyped buffer with a particular alignment requirement on the pointer I have no way of saying that to RMM.

**Describe the solution you'd like**

Add an alignment parameter.

**Describe alternatives you've considered**

Roll my own buffer.

**Additional context**

- #2174 covers alignment requirements in device_uvector.

### PR DESCRIPTION
**Pull request by wence-**
## Description

Allow construction of a `device_buffer` (and thence `device_uvector`) where we explicitly specify the required alignment of the allocation.

All device MRs currently throw if requesting an alignment of more than 256 bytes, even the aligned resource adaptor, because they all go through the `do_allocate` implementation in `device_memory_resource` that throws. Since this route is going away soon I am not inclined to fix that here. Going forward, all "leaf" memory resources should explicitly throw if they cannot return allocations satisfying the requested alignment.

- Closes #2329
- Partially addresses #2174 by passing through the required alignment of the type to the storage allocation. This will now explicitly throw if the MR cannot satisfy the requested alignment.

This is an **ABI breaking** change.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/rmm/device_uvector.hpp [hunk 1]
```diff
@@ -117,14 +117,17 @@ class device_uvector {
...
+   *
    * @param size The number of elements to allocate storage for
    * @param stream The stream on which to perform the allocation
    * @param mr The resource used to allocate the device storage
    */
   explicit device_uvector(size_type size,
                           cuda_stream_view stream,
                           device_async_resource_ref mr = mr::get_current_device_resource_ref())
-    : _storage{elements_to_bytes(size), stream, mr}
+    : _storage{elements_to_bytes(size), std::alignment_of_v<T>, stream, mr}
```
Comment (bdice) [inline cpp/include/rmm/device_uvector.hpp:130]: Should this be `cuda::std::alignment_of_v` since we're dealing with device data? I don't have a good grasp on whether we should prefer `cuda::std::` in this case or not.

File: cpp/include/rmm/error.hpp [hunk 2]
```diff
@@ -100,4 +100,13 @@ class out_of_range : public std::out_of_range {
   using std::out_of_range::out_of_range;
 };
 
+/**
+ * @brief Exception thrown when an argument to a function is invalid
+ *
+ * @ingroup errors
+ */
+class invalid_argument : public std::invalid_argument {
```
Comment (bdice) [inline cpp/include/rmm/error.hpp:108]: Do you think it is useful (or good practice) to define a custom exception here rather than using the `std::invalid_argument` directly? I am inclined to avoid adding new custom exception types if they don't provide significant value.

File: cpp/tests/device_buffer_tests.cu [hunk 3]
```diff
@@ -509,3 +511,116 @@ TYPED_TEST(DeviceBufferTest, SetGetStream)
...
+  rmm::device_buffer buff(100, alignment, rmm::cuda_stream_default);
+  EXPECT_EQ(buff.alignment(), alignment);
+  EXPECT_TRUE(rmm::is_pointer_aligned(buff.data(), alignment));
+  EXPECT_EQ(buff.size(), 100);
+}
+
+TEST(DeviceBufferAlignmentTest, ExplicitAlignmentTooLarge)
+{
+  auto constexpr alignment = rmm::CUDA_ALLOCATION_ALIGNMENT * 2;
+  EXPECT_THROW(rmm::device_buffer(100, alignment, rmm::cuda_stream_default), rmm::bad_alloc);
```
Comment (bdice) [inline cpp/tests/device_buffer_tests.cu:542]: My plan for handling this in the future: once we merge `staging` into `main`, `device_buffer` will own its memory resource. That means we could automatically wrap the provided MR with an `aligned_resource_adaptor` and use that for allocation when we are asked for higher alignment than `rmm::CUDA_ALLOCATION_ALIGNMENT`.

What do you think? Does wrapping the user MR with an aligned adaptor like this violate the user expectation that the `device_buffer` uses the "same MR" that was passed in? Otherwise we would probably need to adapt the machinery of every MR to properly support alignment...?

File: cpp/tests/device_buffer_tests.cu [hunk 4]
```diff
@@ -509,3 +511,116 @@ TYPED_TEST(DeviceBufferTest, SetGetStream)
...
+  EXPECT_EQ(buff.alignment(), alignment);
+  EXPECT_EQ(buff.size(), 0);
+  EXPECT_TRUE(buff.is_empty());
+  EXPECT_TRUE(rmm::is_pointer_aligned(buff.data(), alignment));
+}
+
+TEST(DeviceBufferAlignmentTest, EmptyBufferAlignmentTooLarge)
+{
+  auto constexpr alignment = rmm::CUDA_ALLOCATION_ALIGNMENT * 2;
+  rmm::device_buffer buff(std::size_t{0}, alignment, rmm::cuda_stream_default);
```
Comment (bdice) [inline cpp/tests/device_buffer_tests.cu:622]: I think this should throw. We're constructing with an unsupported (too large) alignment. We shouldn't permit this with a zero-sized allocation if we disallow it with a nonzero-sized allocation.

File: cpp/tests/device_buffer_tests.cu [hunk 5]
```diff
@@ -509,3 +511,116 @@ TYPED_TEST(DeviceBufferTest, SetGetStream)
...
+  EXPECT_EQ(buff.alignment(), alignment);
+  EXPECT_TRUE(buff.is_empty());
+  EXPECT_TRUE(rmm::is_pointer_aligned(buff.data(), alignment));
```
Comment (bdice) [inline cpp/tests/device_buffer_tests.cu:625]: Does this mean `nullptr` is aligned to any requested alignment? I would've expected `false` naively but it appears to be `true`.

https://godbolt.org/z/zYra1qWq8

File: cpp/tests/device_buffer_tests.cu (hunk 4 above)
Comment (wence-) [reply cpp/tests/device_buffer_tests.cu:622]: The postcondition for alignment is `(uintptr_t)ptr % alignment == 0`, this postcondition is always true if `ptr` is nullptr.

File: cpp/tests/device_buffer_tests.cu (hunk 5 above)
Comment (wence-) [reply cpp/tests/device_buffer_tests.cu:625]: Yes.

File: cpp/include/rmm/error.hpp (hunk 2 above)
Comment (wence-) [reply cpp/include/rmm/error.hpp:108]: I was copy-pasta-ing from the existing exception types in this file. e.g. we already forward `std::out_of_range`.

File: cpp/include/rmm/device_uvector.hpp (hunk 1 above)
Comment (wence-) [reply cpp/include/rmm/device_uvector.hpp:130]: No, the type's alignment does not depend on whether you're dealing with device data or not. We would use `cuda::std::` if we needed this ctor to be `__host__ __device__`, but we can't possibly need that.

File: cpp/tests/device_buffer_tests.cu (hunk 3 above)
Comment (wence-) [reply cpp/tests/device_buffer_tests.cu:542]: I think that is a reasonable thing to do. I would like if CCCL resources supported arbitrary alignments (but they don't today, they also throw if requesting more than 256-byte aligned allocations), I opened https://github.com/NVIDIA/cccl/issues/8157.

Note, as part of that, I think at the very least memory resources should provide a property that allows us to statically query the alignment they provide. That way one could wrap an `aligned` adaptor around any memory resource and not have the hope that it provides 256-byte alignment for "fast-path" purposes (like we do today).

File: cpp/include/rmm/error.hpp (hunk 2 above)
Comment (bdice) [reply cpp/include/rmm/error.hpp:108]: Yeah. That was a historical decision and I don't know the complete rationale. I'm happy to follow your judgment on this. If you don't think it's necessary, I won't require it.

File: cpp/tests/device_buffer_tests.cu (hunk 3 above)
Comment (bdice) [reply cpp/tests/device_buffer_tests.cu:542]: I see. Yes, something like a `minimum_alignment()` guarantee could help.

File: cpp/tests/device_buffer_tests.cu (hunk 4 above)
Comment (bdice) [reply cpp/tests/device_buffer_tests.cu:622]: Imagine you have `HugeAlignedType`. You think it should be possible to construct a `device_uvector<HugeAlignedType>` with zero elements but not possible to construct with one element? In my understanding that is what this would allow.

The "postcondition for alignment" argument makes sense, but I think it's important that we don't allow unsupported alignments in edge cases like zero elements.

Comment (bdice) [reply cpp/tests/device_buffer_tests.cu:622]: This is the only really questionable behavior left in this PR for me, once we come to a resolution here I can approve.

Comment (wence-) [reply cpp/tests/device_buffer_tests.cu:622]: > You think it should be possible to construct a `device_uvector<HugeAlignedType>` with zero elements but not possible to construct with one element? In my understanding that is what this would allow.

Yes, because all the invariants required of the zero-element `device_uvector` are upheld, whereas those of the non-empty one are not.

File: cpp/include/rmm/error.hpp (hunk 2 above)
Comment (wence-) [reply cpp/include/rmm/error.hpp:108]: The usual reason is encapsulation. RMM functions throw some exception that derives from `rmm::exception` then I can distinguish errors in RMM from errors elsewhere.

That said, our encapsulation doesn't do that today: RMM errors are distinguished by their name, not a common base. So I might just remove.

File: cpp/tests/device_buffer_tests.cu (hunk 4 above)
Comment (bdice) [reply cpp/tests/device_buffer_tests.cu:622]: I suppose I wouldn't separate those sets of invariants (zero vs nonzero sizes are the same type and should obey the same constructor rules) but I'll accept this.

**Comment by bdice** [comment]
@wence- I approved, please merge when you're happy. It sounded like you were debating on whether to keep or remove the error class wrapper.

File: cpp/include/rmm/error.hpp (hunk 2 above)
Comment (wence-) [reply cpp/include/rmm/error.hpp:108]: Decided to leave it....
